### PR TITLE
25124 enforcing password requirements

### PR DESCRIPTION
**Pull request by Serp1co**
## What does this PR change?

Enforce password requirement for server users, comes with a new admin panel to manage local users password policy.

## GUI diff

Before:
uyuni-beofre

After:
uyuni-after
image

- [x] **DONE**

## Documentation
- Documentation issue was created: [Link for SUSE Manager contributors](https://github.com/SUSE/spacewalk/issues/new?template=ISSUE_TEMPLATE_DOCUMENTATION.md&labels=documentation&projects=SUSE/spacewalk/31), [Link for community contributors](https://github.com/uyuni-project/uyuni-docs/issues/new).
- API documentation added: please review the Wiki page [Writing Documentation for the API](https://github.com/uyuni-project/uyuni/wiki/Writing-documentation-for-the-API) if you have any changes to API documentation.
- (OPTIONAL) [Documentation PR](https://github.com/uyuni-project/uyuni-docs/pulls)

- [ ] **DONE**

## Test coverage
- Unit tests were added

- [ ] **DONE**

## Links

Issue(s): https://github.com/SUSE/spacewalk/issues/25124
Port(s): # **add downstream PR(s), if any**

- [ ] **DONE**

## Changelogs

Make sure the changelogs entries you are adding are compliant with https://github.com/uyuni-project/uyuni/wiki/Contributing#changelogs and https://github.com/uyuni-project/uyuni/wiki/Contributing#uyuni-projectuyuni-repository

If you don't need a changelog check, please mark this checkbox:

- [ ] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [x] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql" (Test skipped, there are no changes to test)
- [ ] Re-run test "java_pgsql_tests"        
- [ ] Re-run test "schema_migration_test_pgsql"      
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests" (Test skipped, there are no changes to test)

# Before you merge

Check [How to branch and merge properly](https://github.com/uyuni-project/uyuni/wiki/How-to-branch-and-merge-properly)!

### CODE REVIEW

File: web/html/src/manager/admin/password-policy/password-policy.tsx [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+      isUyuni: props.isUyuni,
+    },
+    errors: {},
+  });
+
+  const { updatePolicy } = usePasswordPolicyApi();
+
+  return (
+    <TopPanel
+      title={t(policy.properties.isUyuni ? "SUSE Manager" : "Uyuni").concat(" Configuration - Password Policy")}
```
Comment (Etheryte) [inline web/html/src/manager/admin/password-policy/password-policy.tsx:51]: This won't work for our translators, because the order of the words will be different in different languages. It's a bit more verbose, but the below would be preferred from the translation perspective:  

```tsx
title={policy.properties.isUyuni ? t("...") : t("...")}
```

Alternatively, we could sidestep the whole problem by not including the product name here at all, e.g.

```tsx
title={t("Password Policy Configuration")}
```

Personally I prefer the second approach, but I don't have a strong preference.

File: web/html/src/manager/admin/password-policy/password-policy.tsx [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+
+import withPageWrapper from "components/general/with-page-wrapper";
+import { Form, Text, Check } from "components/input";
+import { Panel } from "components/panels/Panel";
+import { TopPanel } from "components/panels/TopPanel";
+import { AsyncButton } from "components/buttons";
+import { showErrorToastr, showSuccessToastr } from "components/toastr/toastr";
+
+// Assume you have an API hook similar to useLifecyclePaygActionsApi
+import usePasswordPolicyApi from "manager/admin/password-policy/use-password-policy-api";
```
Comment (Etheryte) [inline web/html/src/manager/admin/password-policy/password-policy.tsx:11]: This file does not seem to exist in this pull request, did you perhaps forget to `git add` it?

File: java/code/src/com/redhat/rhn/common/util/validation/password/PasswordPolicy.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
+package com.redhat.rhn.common.util.validation.password;
```
Comment (rjmateus) [inline java/code/src/com/redhat/rhn/common/util/validation/password/PasswordPolicy.java:1]: I think we are missing the license comment block in here.

File: java/code/src/com/redhat/rhn/common/util/validation/password/PasswordPolicyCheckFail.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
+package com.redhat.rhn.common.util.validation.password;
```
Comment (rjmateus) [inline java/code/src/com/redhat/rhn/common/util/validation/password/PasswordPolicyCheckFail.java:1]: license block is missing

File: java/code/src/com/redhat/rhn/common/util/validation/password/PasswordValidationException.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
+package com.redhat.rhn.common.util.validation.password;
```
Comment (rjmateus) [inline java/code/src/com/redhat/rhn/common/util/validation/password/PasswordValidationException.java:1]: license block missing

**Comment by Etheryte** [comment]
The frontend parts that are currently there look good. 👍

I assume there will be followup work to fix the build etc.

File: java/code/src/com/suse/manager/webui/controllers/admin/handlers/PasswordPolicyController.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+     * @return an OK string
+     */
+    public String postPasswordPolicy(Request request, Response response, User user) {
+        if (!user.hasRole(SAT_ADMIN)) {
+            throw new PermissionException(SAT_ADMIN);
+        }
+        PasswordPolicyProperties passwordPolicyProperties = GSON.fromJson(request.body(),
+                PasswordPolicyProperties.class);
+        try {
+            SatConfigFactory.setSatConfigValue(SatConfigFactory.PSW_CHECK_LENGTH_MIN,
```
Comment (rjmateus) [inline java/code/src/com/suse/manager/webui/controllers/admin/handlers/PasswordPolicyController.java:67]: Should we validate if the user enters something crazy in here? Like a negative max length? Or a min length bigger then max length?

**Review by rjmateus — COMMENTED** [review]
Some small notes

File: java/code/src/com/suse/manager/webui/controllers/admin/handlers/PasswordPolicyController.java (hunk 6 above)
Comment (Serp1co) [reply java/code/src/com/suse/manager/webui/controllers/admin/handlers/PasswordPolicyController.java:67]: yeah we should, i didn't think about that. Also for better ux we should enforce this rules in the frontend.

File: java/code/src/com/suse/manager/webui/controllers/admin/beans/PasswordPolicyProperties.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) 2021 SUSE LLC
```
Comment (mcalmer) [inline java/code/src/com/suse/manager/webui/controllers/admin/beans/PasswordPolicyProperties.java:2]:
```suggestion
 * Copyright (c) 2024 SUSE LLC
```

File: schema/spacewalk/common/data/rhnConfiguration.sql [hunk 8]
```diff
@@ -3,3 +3,23 @@ INSERT INTO rhnConfiguration (key, description, default_value) VALUES ('extauth_
...
+INSERT INTO rhnConfiguration (key, description, value, default_value)
+VALUES ('password_check_consecutive_char_flag', 'Password has to have no consecutive characters', FALSE, FALSE);
+INSERT INTO rhnConfiguration (key, description, value, default_value)
+VALUES ('password_check_special_char_flag', 'Password has to have at least a special character', FALSE, FALSE);
+INSERT INTO rhnConfiguration (key, description, value, default_value)
+VALUES ('password_check_restricted_occurrence_flag', 'Password has to have no repeating characters', FALSE, FALSE);
+INSERT INTO rhnConfiguration (key, description, value, default_value)
+VALUES ('password_check_max_occurrence', 'Maximum number of valid occurrence of a character', 2, 2);
+INSERT INTO rhnConfiguration (key, description, value, default_value)
+VALUES ('password_check_special_characters', 'List of special characters to check in a password', null, '!$%&()*+,./:;<=>?[\\]^_{|}~');
```
Comment (mcalmer) [inline schema/spacewalk/common/data/rhnConfiguration.sql:25]: We need to think about the defaults. What policy do we want to configure as default?
Maybe something what NIST, BSI & Co suggest?

**Comment by mcalmer** [comment]
@Serp1co You need to implement DB schema migration and also think about how to migrate from current configuration file values to DB. We have at least min and max password length. If they are changed, we should set it the DB configuration

File: java/code/src/com/suse/manager/webui/controllers/admin/handlers/PasswordPolicyController.java (hunk 6 above)
Comment (rjmateus) [reply java/code/src/com/suse/manager/webui/controllers/admin/handlers/PasswordPolicyController.java:67]: We can check that in the front-end, but all validations must also be done in the backend also.

**Review by cbbayburt — COMMENTED** [review]
Once the default values are decided, we need a button in the UI to reset everything to default. If you want to go one step further for convenience, we can decide on a few sensible presets with varying levels of restriction, and add buttons for each, which set all the fields accordingly.

Ideally, all those individual fields must be filled only if the user has some specific rules in mind. Otherwise, we should offer shortcuts like I mentioned above.

File: java/code/src/com/suse/manager/webui/controllers/admin/handlers/PasswordPolicyController.java [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    /**
+     * edit the current password policy settings
+     * @param request
+     * @param response
+     * @param user
+     * @return an OK string
+     */
+    public String postPasswordPolicy(Request request, Response response, User user) {
```
Comment (mcalmer) [inline java/code/src/com/suse/manager/webui/controllers/admin/handlers/PasswordPolicyController.java:60]: We also need an API for the user (XMLRCP/JSON over HTTP) to get and modify the policy

**Comment by rjmateus** [comment]
@cbbayburt what about showing the default value in front, maybe in gray, so customers know what was the default value, and if they want come back to that value for a few fields?

**Comment by cbbayburt** [comment]
> @cbbayburt what about showing the default value in front, maybe in gray, so customers know what was the default value, and if they want come back to that value for a few fields?

I'm not sure. Especially the "special chars" field is not so convenient to type in manually. Maybe @bisht-richa or @Etheryte has some ideas?

**Comment by bisht-richa** [comment]
> > @cbbayburt what about showing the default value in front, maybe in gray, so customers know what was the default value, and if they want come back to that value for a few fields?
> 
> I'm not sure. Especially the "special chars" field is not so convenient to type in manually. Maybe @bisht-richa or @Etheryte has some ideas?

Some rough idea 

Screenshot 2024-11-05 at 16 32 03

File: web/html/src/manager/admin/password-policy/password-policy.tsx (hunk 1 above)
Comment (Serp1co) [reply web/html/src/manager/admin/password-policy/password-policy.tsx]: Agreed with the 2nd approach

File: schema/spacewalk/common/data/rhnConfiguration.sql (hunk 8 above)
Comment (Serp1co) [reply schema/spacewalk/common/data/rhnConfiguration.sql]: I think we should keep the old defaults, as agreed with @admd, so that we don't break the old tests and the automations already in place

Comment (Serp1co) [reply schema/spacewalk/common/data/rhnConfiguration.sql]: Defaulted with previous values for compatibility reason as agreed with @admd

File: java/code/src/com/suse/manager/webui/controllers/admin/handlers/PasswordPolicyController.java (hunk 6 above)
Comment (Serp1co) [reply java/code/src/com/suse/manager/webui/controllers/admin/handlers/PasswordPolicyController.java]: validation implemented for values corner cases

**Comment by mcalmer** [comment]
@Serp1co there is one cucumber integration test which fail and the error seems to happen in a password change test. So maybe caused by this PR.

**Comment by bisht-richa** [comment]
UI looks good. Just two things:
- Align buttons to the form elements column 
- Only `Save` button should be **btn-primary** please change `Reset` button to **

Screenshot 2025-01-13 at 11 40 08**

File: java/code/src/com/redhat/rhn/common/conf/UserDefaults.java [hunk 10]
```diff
@@ -20,8 +20,6 @@
  */
 public class UserDefaults {
     private static UserDefaults instance = new UserDefaults();
-    public static final String MAX_PASSWORD_LENGTH = "max_passwd_len";
-    public static final String MIN_PASSWORD_LENGTH = "min_passwd_len";
```
Comment (mcalmer) [inline java/code/src/com/redhat/rhn/common/conf/UserDefaults.java:24]: You should remove these configurations also from the "default config".
You find it in `python/spacewalk/rhn-conf/rhn_server.conf`

**Review by mcalmer — CHANGES_REQUESTED** [review]
I also found 2 cases of using "getConfig('max_passwd_len')"
- java/code/src/com/suse/manager/webui/controllers/login/LoginController.java
- java/code/webapp/WEB-INF/pages/common/fragments/user/user_attribute_sizes.jspf

In the last file also "min_passwd_len" is used as well.

File: java/code/src/com/redhat/rhn/domain/common/RhnConfigurationFactory.java [hunk 11]
```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright (c) 2014 Red Hat, Inc.
```
Comment (mcalmer) [inline java/code/src/com/redhat/rhn/domain/common/RhnConfigurationFactory.java:2]:
```suggestion
 * Copyright (c) 2025 SUSE LLC
```

**Review by mcalmer — APPROVED** [review]
One suggestion about to change a header. But the rest of the java and DB changes looks good

File: spacewalk/admin/uyuni-update-config [hunk 12]
```diff
@@ -1,89 +1,176 @@
...
+        if CFG.has_key("server.min_passwd_len"):
+            move_configuration(
+                "PSW_CHECK_LENGTH_MIN",
+                CFG.max_passwd_len,
+                oldkey="server.min_passwd_len",
+            )
+        elif CFG.has_key("min_passwd_len"):
+            move_configuration(
+                "PSW_CHECK_LENGTH_MIN", CFG.min_passwd_len, oldkey="min_passwd_len"
+            )
```
Comment (srbarrios) [inline spacewalk/admin/uyuni-update-config:73]: To be reviewed.
I think we need to use it on uppercase, so CFG.MIN_PASSWD_LEN
